src/Spectre.libGaussianMixtureModelling/ExpectationRunnerOpt.h

```diff
@@ -0,0 +1,113 @@
+namespace spectre::unsupervised::gmm
+{
+    /// <summary>
+    /// Class serves the purpose of expectation step of Expectation
+    /// Maximization algorithm. It is partially optimized, but still uses
+    /// just a single thread to compute.
+    /// </summary>
+    class ExpectationRunnerOpt
+    {
+    public:
+        /// <summary>
+        /// Constructor initializing the class with data required during
+        /// expectation step.
+        /// </summary>
+        /// <param name="spectrum">Spectrum to operate on.</param>
+        /// <param name="affilationMatrix">Matrix symbolising the probability
+        /// of affilation of each sample to a certain gaussian component.
+        /// </param>
+        /// <param name="components">Gaussian components.</param>
+        ExpectationRunnerOpt(SpectrumView spectrum,
+            Matrix<DataType> &affilationMatrix,
+            std::vector<GaussianComponent> &components)
+            : m_Spectrum(spectrum), m_Components(components)
+            , m_AffilationMatrix(affilationMatrix)
+            , m_Sums(spectrum.mzs.size())
+        {
+        }
+
+        /// <summary>
+        /// Fills affilation (gamma) matrix with probabilities of affilation
+        /// of each sample to a certain gaussian component.
+        /// </summary>
+        void Expectation()
+        {
+            DataView mzs = m_Spectrum.mzs;
+            DataView intensities = m_Spectrum.intensities;
+            std::fill(m_Sums.begin(), m_Sums.end(), 0.0);
+            const unsigned numberOfComponents = (unsigned)m_Components.size();
+            for (unsigned k = 0; k < numberOfComponents - 1; k++)
+            {
+                for (unsigned i = 0; i < (unsigned)mzs.size(); i++)
+                {
+                    DataType probability = m_Components[k].weight *
+                        Gaussian(mzs[i], m_Components[k].mean,
+                            m_Components[k].deviation);
+                    m_AffilationMatrix[k][i] = probability;
+                    m_Sums[i] += probability;
+                }
+            }
+            const unsigned lastK = numberOfComponents - 1;
+            for (unsigned i = 0; i < (unsigned)mzs.size(); i++)
+            {
+                DataType probability = m_Components[lastK].weight *
+                    Gaussian(mzs[i], m_Components[lastK].mean,
+                        m_Components[lastK].deviation);
+                m_AffilationMatrix[lastK][i] = probability;
+                m_Sums[i] = 1.0 / (m_Sums[i] + probability);
+            }
+
+            for (unsigned k = 0; k < numberOfComponents; k++)
+            {
+                for (unsigned i = 0; i < (unsigned)mzs.size(); i++)
+                {
+                    m_AffilationMatrix[k][i] *= m_Sums[i] * intensities[i];
+                }
+            }
+        }
+
+    private:
+        SpectrumView m_Spectrum;
+        Matrix<DataType> &m_AffilationMatrix;
+        std::vector<GaussianComponent> &m_Components;
```

Comment:
Who is the owner of this data? What if the owner gets deleted?

Reply:
Owner of the data is the function using the class. I don't like this design either, but I had to implement it this way in order to conform with exisiting API which assumed storing the references to matrix and components upon creation. This will be refactored in the future.

Reply:
I think that passing these as arguments into `Expectation` function is a way to go.

Reply:
If design is bad, fix it right now. We do not want to merge poorly designed code. What is purpose of CR then? @Daroslaw is right in this case - this should be an argument into `Expectation` function or any other way around, not like this.

You cannot ensure from this point, that owner of this data does not get deleted. Situation you described is just one of possible usages. You do not prevent any other usage.

It's poorly designed, if correctness does not follow from design.

I'd accept it, when there is no other way. But there is.